selfdrive/car/tesla/radar_interface.py

```diff
@@ -1,5 +1,113 @@
 #!/usr/bin/env python3
+from cereal import car
+from opendbc.can.parser import CANParser
+from selfdrive.car.tesla.values import DBC, CANBUS
 from selfdrive.car.interfaces import RadarInterfaceBase
 
+RADAR_MSGS_A = list(range(0x371, 0x37E, 3))
+RADAR_MSGS_B = list(range(0x372, 0x37F, 3))
+NUM_POINTS = len(RADAR_MSGS_A)
+SGU_INDEX_VALUES = [f"RADC_ACCTargObj{i+1}_sguIndex" for i in range(NUM_POINTS)]
+SGU_INDEX_NONE = 63
+
+def get_radar_can_parser(CP):
+  # Status messages
+  signals = [
+    ('RADC_HWFail', 'TeslaRadarSguInfo', 0),
+    ('RADC_SGUFail', 'TeslaRadarSguInfo', 0),
+    ('RADC_SensorDirty', 'TeslaRadarSguInfo', 0),
+  ]
+
+  checks = [
+    ('TeslaRadarSguInfo', 10),
+    ('TeslaRadarTguInfo', 10),
+  ]
+
+  # Radar tracks. There are also raw point clouds available,
+  # we don't use those.
+  for i in range(NUM_POINTS):
+    msg_id_a = RADAR_MSGS_A[i]
+    msg_id_b = RADAR_MSGS_B[i]
+
+    # There is a bunch more info in the messages,
+    # but these are the only things actually used in openpilot
+    signals.extend([
+      ('LongDist', msg_id_a, 255),
+      ('LongSpeed', msg_id_a, 0),
+      ('LatDist', msg_id_a, 0),
+      ('LongAccel', msg_id_a, 0),
+      ('Meas', msg_id_a, 0),
+      ('Tracked', msg_id_a, 0),
+      ('Index', msg_id_a, 0),
+
+      ('LatSpeed', msg_id_b, 0),
+      ('Index2', msg_id_b, 0),
+
+      (SGU_INDEX_VALUES[i], "TeslaRadarTguInfo", 63),
+    ])
+
+    checks.extend([
+      (msg_id_a, 8),
+      (msg_id_b, 8),
+    ])
+
+  return CANParser(DBC[CP.carFingerprint]['radar'], signals, checks, CANBUS.radar)
+
 class RadarInterface(RadarInterfaceBase):
-  pass
+  def __init__(self, CP):
+    super().__init__(CP)
+    self.rcp = get_radar_can_parser(CP)
+    self.updated_messages = set()
+
+  def update(self, can_strings):
+    if self.rcp is None:
+      return super().update(None)
+
+    values = self.rcp.update_strings(can_strings)
+    self.updated_messages.update(values)
+
+    # Trigger update only on new first track message
+    if RADAR_MSGS_A[0] not in self.updated_messages:
```

Comment:
Why not on the last message? Maybe use the same structure as other radar interfaces with `self.trigger_message`

Reply:
didn't really see the benefit of using `self.trigger_message`, since it's just another way to specify the last message (most often hard-coded in the other car ports, doesn't seem very clean).
I'll change it to `RADAR_MSGS_B[-1]`

Reply:
not sure if they always get sent out in sequence, but it does make more intuitive sense to use the last one instead of the first one I guess

Reply:
just noticed that this trigger doesn't actually do anything (or at least not what I imagined it did), since `self.updated_messages` is not cleared, so this is always true

Reply:
Isn't the trigger message field used by process replay?